In [1]:
import pandas as pd
 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

df.head(5)
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [ ]:
##A##

import pandas as pd 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

print(df['Survived'].value_counts(normalize = True))

Survived
0    0.616162
1    0.383838
Name: proportion, dtype: float64


1. ¿Cuál es la tasa de supervivencia global del barco expresada en porcentaje?

38.38%

In [ ]:
##B##

import pandas as pd 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

print("Promedio de la columna 'Survived': ", df['Survived'].mean())
print("Promedio de mujeres sobrevivientes: ", df[df['Sex'] == 'female']['Survived'].mean())
print("Promedio de hombres sobrevivientes: ", df[df['Sex'] == 'male']['Survived'].mean())

Promedio de la columna 'Survived':  0.3838383838383838
Promedio de mujeres sobrevivientes:  0.7420382165605095
Promedio de hombres sobrevivientes:  0.18890814558058924


1. ¿Qué porcentaje exacto de mujeres sobrevivió en contraste con los hombres?

El porcentaje de mujeres fue de 74.20%, mientras que el de hombres fue de 18.89%

In [ ]:
##C##

import pandas as pd 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

Q1 = df['Fare'].quantile(0.25)
Q3 = df['Fare'].quantile(0.75)
IQR = Q3 - Q1
limite = Q3 + 1.5 * IQR

print("Q1:", Q1, "| Q3:", Q3, "| IQR:", IQR, "| Límite superior:", limite, "| Cantidad de outliers:", len(df[df['Fare'] > limite]))

print(df[df['Fare'] > limite]['Pclass'].value_counts())

Q1: 7.9104 | Q3: 31.0 | IQR: 23.0896 | Límite superior: 65.6344 | Cantidad de outliers: 116
Pclass
1    104
3      7
2      5
Name: count, dtype: int64


1. ¿A qué clase (Pclass) pertenecían la mayoría de ellos?

La mayoria de los pasajeros que pagaron una tarifa considerada outlier pertenecian a la primera clase.

In [11]:
##D##

import pandas as pd 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

print("Mediana de la columna 'Fare': ", df['Fare'].median())
print("Promedio de la columna 'Fare': ", df['Fare'].mean())

Mediana de la columna 'Fare':  14.4542
Promedio de la columna 'Fare':  32.204207968574636


1. ¿Qué significa que la media sea tan superior a la mediana?
Significa que existen algunas tarifas muy altas que aumentan el promedio

2. ¿Cómo afectará esta asimetría al aprendizaje del modelo?
Si no se escala adecuadamente la variable, las tarifas altas podrian tener demasiado peso en las distancias y afectar el aprendizaje del modelo

In [10]:
##E##

import pandas as pd 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

sobrevivientes = df[df['Survived'] == 1].sample(n=58, random_state=42)
no_sobrevivientes = df[df['Survived'] == 0].sample(n=92, random_state=42)

muestra = pd.concat([sobrevivientes, no_sobrevivientes])

print("Sobrevivientes:", len(sobrevivientes), "| No sobrevivientes:", len(no_sobrevivientes), "| Total:", len(muestra))

Sobrevivientes: 58 | No sobrevivientes: 92 | Total: 150


1. ¿Qué sesgo evitan al hacer esto?
Teniendo en cuenta que tenemos 58 sobrevivientes y 92 no sobrevivientes, podemos conocer que una muestra aleatoria podría no representar correctamente la distribución de la variable Survived

In [ ]:
##F##

import pandas as pd 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

print(df.groupby('Survived')['Age'].mean())

Survived
0    30.626179
1    28.343690
Name: Age, dtype: float64


1. ¿Qué sesgo estadístico estamos introduciendo involuntariamente en el resultado de esa función y cómo afectaría la inferencia de nuestro modelo?

Se esta introduciendo un sesgo por datos faltantes, porque el import de pandas ignora los valores NaN al calcular la media, lo que podría afectar las conclusiones y hacer que el modelo aprenda una relación incorrecta entre la edad y la supervivencia.

In [ ]:
##G##

import pandas as pd 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

print("Tarifa máxima:", df['Fare'].max(), "| Límite superior:", limite, "| Outliers:", len(df[df['Fare'] > limite]))

Tarifa máxima: 512.3292 | Límite superior: 65.6344 | Outliers: 116


1. ¿Deberíamos eliminar estas filas con .drop() antes de entrenar nuestro modelo?

No los eliminaria automaticamente, porque representan casos reales y contienen información importante sobre la clase y las condiciones del pasajero, por lo tanto, estos son útiles para comprobar todo.

In [ ]:
##H##

import pandas as pd 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

df['Titulo'] = df['Name'].str.extract(r',\s*([^.]*)\.')

print(df.groupby('Titulo')['Age'].median())

Titulo
Capt            70.0
Col             58.0
Don             40.0
Dr              46.5
Jonkheer        38.0
Lady            48.0
Major           48.5
Master           3.5
Miss            21.0
Mlle            24.0
Mme             24.0
Mr              30.0
Mrs             35.0
Ms              28.0
Rev             46.5
Sir             49.0
the Countess    33.0
Name: Age, dtype: float64


1. ¿Por qué imputar las edades faltantes basándose en la mediana del "Título" (ej. "Master" = niño, "Mr" = adulto) sería estadísticamente superior y reduciría el error de nuestro futuro modelo, en comparación con usar la mediana global?

Utilizar la mediana de edad según el título sería mejor que utilizar una sola mediana global porque los títulos proporcionan información sobre el tipo de pasajero.

In [ ]:
##I##

import pandas as pd 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

print("Varianza de Survived:", df['Survived'].var())

Varianza de Survived: 0.2367722165474984


1. ¿Qué significaría si la varianza de esta variable fuera exactamente 0.0? ¿Qué pasaría si intentan entrenar un algoritmo de clasificación con un dataset donde la variable de respuesta tiene varianza 0.0?

La varianza de Survived es aproximadamente 0.237. Si la varianza fuera exactamente de 0.0, significaria que todos los pasajeros tendrian el mismo valor de Survived, es decir, que todos murieron o todos sobrevivieron, en este caso, un algoritmo de clasificación no podría aprender a diferenciar entre dos clases.

In [ ]:
##J##

import pandas as pd 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

print(df.groupby(['Pclass', 'Sex', 'Embarked'])['PassengerId'].count())

Pclass  Sex     Embarked
1       female  C            43
                Q             1
                S            48
        male    C            42
                Q             1
                S            79
2       female  C             7
                Q             2
                S            67
        male    C            10
                Q             1
                S            97
3       female  C            23
                Q            33
                S            88
        male    C            43
                Q            39
                S           265
Name: PassengerId, dtype: int64


1. ¿Qué fenómeno perjudicial (sobreajuste o subajuste) ocurrirá inevitablemente si dejamos que el modelo aprenda reglas basadas en esos grupos de 1 solo pasajero?

Los grupos que contienen solamente uno o pocos pasajeros pueden provocar sobreajuste. El modelo podría aprender reglas demasiado específicas basándose en casos individuales, en lugar de encontrar patrones generales.